lib/Parser.js

```diff
@@ -0,0 +1,81 @@
+const fs = require('fs-extra');
+
+// First char: https://github.com/elm/compiler/blob/2860c2e5306cb7093ba28ac7624e8f9eb8cbc867/compiler/src/Parse/Variable.hs#L296-L300
+// Rest: https://github.com/elm/compiler/blob/2860c2e5306cb7093ba28ac7624e8f9eb8cbc867/compiler/src/Parse/Variable.hs#L328-L335
+// https://hackage.haskell.org/package/base-4.14.0.0/docs/Data-Char.html#v:isLetter
+const nonCapitalizedNamePart = /\p{Ll}[_\d\p{L}]*/u;
```

Comment:
@harrysarson Here’s how to match an Elm variable name.

Now that I think of it, this also matches reserved names such as `module`, `port` and `if`. I should detect that so the user doesn’t get a really bad error message if the do `port = 3000` by mistake.

Reply:
I tried it out.

Main.elm:

```elm
module Main exposing (..)

import Bad

x = Bad.port
```

Bad.elm:

```
module Bad exposing (..)

port = 5
```

Result:

```
❯ elm make src/Main.elm
-- PORT PROBLEM ---------------------------------------------------- src/Bad.elm

I just saw the start of a `port` declaration, but then I got stuck here:

3| port = 5
        ^
I was expecting to see a name like send or receive next. Something that starts
with a lower-case letter.

Note: Here are some example `port` declarations for reference:

    port send : String -> Cmd msg
    port receive : (String -> msg) -> Sub msg

The first line defines a `send` port so you can send strings out to JavaScript.
Maybe you send them on a WebSocket or put them into IndexedDB. The second line
defines a `receive` port so you can receive strings from JavaScript. Maybe you
get receive messages when new WebSocket messages come in or when the IndexedDB
is changed for some external reason.
```

So it might not be necessary to detect reserved words – Elm seems to look for errors in imports first so the good errors come first. But that’s a bit of an implementation detail to rely on.

Reply:
Is this too evil https://mothereff.in/regexpu#input=const+nonCapitalizedNamePart+%3D+%2F%5Cp%7BLl%7D%5B_%5Cd%5Cp%7BL%7D%5D*%2Fu%3B&dotAllFlag=1&unicodePropertyEscape=1?

Reply:
Yeah, we _could_ do that if we want to support Node.js 8. We could also use `[a-z][_\da-zA-Z]*` for Node.js 8 and call it a day.